tun2socks/lwip/packet_pipe.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lwip
+
+import (
+	"errors"
+	"io"
+	"net"
+	"sync"
+)
+
+// A thread safe in-memory IP packet buffer for LwIPTun2SocksDevice's read and write.
+// Keep in mind there might be multiple reads and writes happending at the same time.
+type packetPipe struct {
+	mu   sync.Mutex
+	once sync.Once
+	eof  bool
+
+	// read buffer (when no WriteTo has been called)
+	mtu int
+	buf net.Buffers
+	rd  chan int
+
+	// WriteTo writer (when an active WriteTo is in-progress)
+	wr      io.Writer
+	wrDone  chan struct{}
+	wrBytes int64
+	wrErr   error
+}
+
+func newPacketPipe(mtu int) *packetPipe {
+	return &packetPipe{
+		mtu: mtu,
+		rd:  make(chan int),
+	}
+}
+
+func (b *packetPipe) close() {
+	b.once.Do(func() {
+		b.mu.Lock()
+		defer b.mu.Unlock()
+		b.eof = true
+		close(b.rd)
+		if b.wr != nil {
+			close(b.wrDone)
+		}
+	})
+}
+
+// read tries to peel one packet from the internal `buf`. If len(buf) == 0, read will block until next packet arrives.
+func (b *packetPipe) read(p []byte) (int, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.eof {
+		return 0, io.EOF
+	}
+
+	// wait until we have at least one valid data in the buffer (blocking)
+	for len(b.buf) == 0 {
+		b.mu.Unlock()
+		<-b.rd
+		b.mu.Lock()
+
+		if b.eof {
+			return 0, io.EOF
+		}
+	}
+
+	if len(p) < len(b.buf[0]) {
+		return 0, io.ErrShortBuffer
```

Comment:
We may want to still write the packet in case of short buffer and return n, because that's what the net package does for UDP. It gives an option to the caller on what to do.